### ISSUE TITLE
service/smart_next_test flaking

### ISSUE THREAD

**Issue by rmacnak-google** [dart-lang/sdk#24176]
We sometimes wind up in Frame(_RootZone.runUnary zone.dart:token(8932)), which would happen if we made a synchronous step where we should have made an asynchronous step. Presumably there is race in getting the latest pause event.

**Comment by turnidge** [comment]
Are the smarts for smart-next in the client or the server?

**Comment by turnidge** [comment]
Potential fix in https://codereview.chromium.org/1313713003

**Comment by turnidge** [comment]
Committed "fix" in 8ed37776d47188036c678a9c4b7f12ac2f248c10.

If we do not see the flaky behavior again soon, we will close this bug.

**Comment by turnidge** [comment]
```
Running smart_next_test.dart [1/27]
Breakpoint reached
Running smart_next_test.dart [2/27]
Checking we are at line 16
We are at Frame(doAsync.<doAsync_async_body> smart_next_test.dart:token(57))
Running smart_next_test.dart [3/27]
next-sync
Running smart_next_test.dart [4/27]
Breakpoint reached
Running smart_next_test.dart [5/27]
Checking we are at line 16
We are at Frame(doAsync.<doAsync_async_body> smart_next_test.dart:token(55))
Running smart_next_test.dart [6/27]
next-async
Running smart_next_test.dart [7/27]
Event: ServiceEvent(owner='isolates/455191489', kind='PauseBreakpoint')
Breakpoint reached
Breakpoint reached
Running smart_next_test.dart [8/27]
Checking we are at line 17
We are at Frame(doAsync.<doAsync_async_body> smart_next_test.dart:token(62))
Running smart_next_test.dart [9/27]
next-sync
Running smart_next_test.dart [10/27]
Breakpoint reached
Running smart_next_test.dart [11/27]
Checking we are at line 18
We are at Frame(doAsync.<doAsync_async_body> smart_next_test.dart:token(67))
Running smart_next_test.dart [12/27]
next-sync
Running smart_next_test.dart [13/27]
Breakpoint reached
Running smart_next_test.dart [14/27]
Checking we are at line 19
We are at Frame(doAsync.<doAsync_async_body> smart_next_test.dart:token(74))
Running smart_next_test.dart [15/27]
next-sync
Running smart_next_test.dart [16/27]
Event: ServiceEvent(owner='isolates/455191489', kind='PauseBreakpoint')
Breakpoint reached
Breakpoint reached
Running smart_next_test.dart [17/27]
Checking we are at line 19
We are at Frame(doAsync.<doAsync_async_body> smart_next_test.dart:token(72))
Running smart_next_test.dart [18/27]
next-async
Running smart_next_test.dart [19/27]
Breakpoint reached
Running smart_next_test.dart [20/27]
Checking we are at line 20
We are at Frame(doAsync.<doAsync_async_body> smart_next_test.dart:token(81))
Running smart_next_test.dart [21/27]
next-sync
Running smart_next_test.dart [22/27]
Event: ServiceEvent(owner='isolates/455191489', kind='PauseBreakpoint')
Breakpoint reached
Running smart_next_test.dart [23/27]
Checking we are at line 20
We are at Frame(doAsync.<doAsync_async_body> smart_next_test.dart:token(79))
Running smart_next_test.dart [24/27]
next-sync
Running smart_next_test.dart [25/27]
Event: ServiceEvent(owner='isolates/455191489', kind='PauseBreakpoint')
Breakpoint reached
Breakpoint reached
Running smart_next_test.dart [26/27]
Checking we are at line 21
We are at Frame(_RootZone.runUnary zone.dart:token(8932))
** Killing script
Unexpected exception in service tests: Expected: <21>
  Actual: <1165>
 #0      DefaultFailureHandler.fail (package:unittest/src/matcher/expect.dart:117:5)
#1      DefaultFailureHandler.failMatch (package:unittest/src/matcher/expect.dart:121:5)
#2      expect (package:unittest/src/matcher/expect.dart:95:20)
#3      stoppedAtLine.<anonymous closure>.<<anonymous closure>_async_body> (file:///Volumes/data/b/build/slave/vm-mac-release-x64-be/build/sdk/runtime/observatory/tests/service/test_helper.dart:214:5)
#4      _asyncThenWrapperHelper.<anonymous closure> (dart:async-patch/async_patch.dart:27)
#5      _rootRunUnary (dart:async/zone.dart:910)
#6      _CustomZone.runUnary (dart:async/zone.dart:810)
#7      _Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:502)
#8      _Future._propagateToListeners (dart:async/future_impl.dart:585)
#9      _Future._completeWithValue (dart:async/future_impl.dart:376)
#10     _Future._asyncComplete.<anonymous closure> (dart:async/future_impl.dart:430)
#11     _rootRun (dart:async/zone.dart:903)
#12     _CustomZone.run (dart:async/zone.dart:802)
#13     _CustomZone.runGuarded (dart:async/zone.dart:708)
#14     _CustomZone.bindCallback.<anonymous closure> (dart:async/zone.dart:733)
#15     _microtaskLoop (dart:async/schedule_microtask.dart:43)
#16     _microtaskLoopEntry (dart:async/schedule_microtask.dart:52)
#17     _runPendingImmediateCallback (dart:isolate-patch/isolate_patch.dart:96)
#18     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:149)



stderr:
Unhandled exception:
Uncaught Error: Expected: <21>
  Actual: <1165>

Stack Trace:
#0      DefaultFailureHandler.fail (package:unittest/src/matcher/expect.dart:117:5)
#1      DefaultFailureHandler.failMatch (package:unittest/src/matcher/expect.dart:121:5)
#2      expect (package:unittest/src/matcher/expect.dart:95:20)
#3      stoppedAtLine.<anonymous closure>.<<anonymous closure>_async_body> (file:///Volumes/data/b/build/slave/vm-mac-release-x64-be/build/sdk/runtime/observatory/tests/service/test_helper.dart:214:5)
#4      _asyncThenWrapperHelper.<anonymous closure> (dart:async-patch/async_patch.dart:27)
#5      _rootRunUnary (dart:async/zone.dart:910)
#6      _CustomZone.runUnary (dart:async/zone.dart:810)
#7      _Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:502)
#8      _Future._propagateToListeners (dart:async/future_impl.dart:585)
#9      _Future._completeWithValue (dart:async/future_impl.dart:376)
#10     _Future._asyncComplete.<anonymous closure> (dart:async/future_impl.dart:430)
#11     _rootRun (dart:async/zone.dart:903)
#12     _CustomZone.run (dart:async/zone.dart:802)
#13     _CustomZone.runGuarded (dart:async/zone.dart:708)
#14     _CustomZone.bindCallback.<anonymous closure> (dart:async/zone.dart:733)
#15     _microtaskLoop (dart:async/schedule_microtask.dart:43)
#16     _microtaskLoopEntry (dart:async/schedule_microtask.dart:52)
#17     _runPendingImmediateCallback (dart:isolate-patch/isolate_patch.dart:96)
#18     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:149)

#0      _rootHandleUncaughtError.<anonymous closure> (dart:async/zone.dart:894)
#1      _microtaskLoop (dart:async/schedule_microtask.dart:43)
#2      _microtaskLoopEntry (dart:async/schedule_microtask.dart:52)
#3      _runPendingImmediateCallback (dart:isolate-patch/isolate_patch.dart:96)
#4      _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:149)


Command[vm]: DART_CONFIGURATION=ReleaseX64 xcodebuild/ReleaseX64/dart --error_on_bad_type --error_on_bad_override --ignore-unrecognized-flags --enable_asserts --enable_type_checks --package-root=xcodebuild/ReleaseX64/packages/ /Volumes/data/b/build/slave/vm-mac-release-x64-be/build/sdk/runtime/observatory/tests/service/smart_next_test.dart
Took 0:00:02.323000

Short reproduction command (experimental):
    python tools/test.py -mrelease -ax64 --write-debug-log --write-test-outcome-log --copy-coredumps --exclude-suite pkg --checked -t60 service/smart_next_test
```

**Comment by turnidge** [comment]
May be fixed by de300cfa0e24aa33559a8625350ae86ca6d1b389.

If the bots look clean, then we will close this bug.

**Comment by rmacnak-google** [comment]
I haven't see any failures on the bots in a few days.
